[GPT2] Add lm_attention_mask as a optional argument

thomasw21 opened the pull request:
# What does this PR do?

We provide a new optional argument in order to pass down a mask that should be applied on attention layers. As the naming conflicts with `attention_mask` (which is use to determine which ones are pad values see https://huggingface.co/transformers/v3.1.0/glossary.html#attention-mask).

Context: we want to be able to obtain a prefixlm like behaviour: Full attention mask on a prefix, and auto regressive masking on suffix.

Some changes that were unrelated to the feature introduction:
 - We've also factorised code from `prepare_inputs_for_generation` into GPT2PretrainedModel as this was duplicated.
 
What this PR doesn't do:
 - Essentially `attention_mask` can we added into `lm_attention_mask` (ie you can put 0's for all pad values and such). However we won't remove `attention_mask` for backward compatibility.
 
TODO:
 - add tests
 
 Example of codes:
 
 ```python
 model = GPT2LMHeadModel.from_pretrained("<MODEL_NAME>")
 model.generate(random_inputs) # autoregressive
 
 lm_attention_mask = torch.ones(1, 1 , random_inputs.shape[1], random_inputs.shape[1], dtype=torch.bool, device=random_inputs.device)
 model.generate(random_inputs, lm_attention_mask=lm_attention_mask) # prefixlm
 
 lm_attention_mask = torch.triu(torch.ones(1, 1 , random_inputs.shape[1], random_inputs.shape[1], dtype=torch.bool, device=random_inputs.device))
 model.generate(random_inputs, lm_attention_mask=lm_attention_mask) # fancy prefixlm
 ```
 
## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

cc @patrickvonplaten @LysandreJik @sgugger

thomasw21 committed:
Add lm_attention_mask as a optional argument in order to pass masked for the attention layers

thomasw21 committed:
Turns out head_mask isn't ambiguous with lm_attention_mask

thomasw21 committed:
Fix non cached version of generate

thomasw21 committed:
make style

thomasw21 committed:
Fix _expand_inputs_for_generation to also expand lm_attention_task if needed

thomasw21 committed:
Woops

thomasw21 committed:
Woops

sgugger reviewed (approved):
Thanks for showing us a prototype. Not a big fan of `lm_attention_mask` as a name, I think I like `language_model_mask` better or `causal_mask` (not sure if this mask might end up in other models than causal LMs though)

thomasw21 committed:
Nit

thomasw21 committed:
Add UT

thomasw21 committed:
Rename lm_attention_mask to attention_block_mask

thomasw21 committed:
make style

thomasw21 committed:
Woops

thomasw21 committed:
Use last_hidden_state attribute

thomasw21 committed:
Woops

thomasw21 committed:
Woops

thomasw21 committed:
Woops

patrickvonplaten commented on src/transformers/generation_utils.py:
```diff
@@ -478,6 +478,11 @@ def _expand_inputs_for_generation(
         if attention_mask is not None:
             model_kwargs["attention_mask"] = attention_mask.index_select(0, expanded_return_idx)
 
+        lm_attention_mask = model_kwargs.get("lm_attention_mask", None)
```
(nit) - would stick to the same signature:

```suggestion
        if "lm_attention_mask" in model_kwargs:
            lm_attention_mask = model_kwargs["lm_attention_mask"]
```

patrickvonplaten commented on src/transformers/generation_utils.py:
```diff
@@ -478,6 +478,11 @@ def _expand_inputs_for_generation(
 
+        lm_attention_mask = model_kwargs.get("lm_attention_mask", None)
+        if lm_attention_mask is not None:
+            if lm_attention_mask.shape[0] != 1:
```
why do we need this? This is also relevant for `batch_size > 1` no?

thomasw21 replied:
So `(lm_attention_mask.shape[0] != 1) == (batch_size > 1)` is should be almost always true. The only time it doesn't match is when we use the same mask for all elements on the match, and thus we allow broadcast for tensors (the same way we will allow broadcast along head dimension). In that specific case, if we can broadcast accross batch_size, we can probably do so along the expanded_batch_size as well.

I'm guessing what you mean is why do I not apply it when `batch_size == 1`? This is also to leverage broadcast as all elements in the expanded_batch should all apply the same mask.

patrickvonplaten commented on src/transformers/models/gpt2/modeling_gpt2.py:
```diff
@@ -487,6 +502,80 @@ def _set_gradient_checkpointing(self, module, value=False):
         if isinstance(module, GPT2Model):
             module.gradient_checkpointing = value
 
+    def prepare_inputs_for_generation(
```
Let's not add this method to all GPT2 models - `GPT2ForTokenClassification` should not be able to call `generate()`

thomasw21 replied:
Though I'm in favor of this option usually, I saw that all `PreTrainedModel` implement `GenerationMixin` which suggest that you could call `generate` with `GPT2ForTokenClassification`.

Would creating a specific generation class `GPT2GenerationMixin` that implements `GenerationMixin` be better? I don't know how this might conflict with GPT2 both implementing `PretrainedModel` and `GPT2GenerationMixin` (which both implement `GenerationMixin`).

patrickvonplaten replied:
It might be a good idea to actually move `GenerationMixin` out of `PreTrainedModel` and to only add it to models that are actually be able to do generate. Happy to discuss this in a new issue.

Why should we though create a `GPT2GenerationMixin` ? This doesn't sound sensible to me. We can simple leave `prepare_inputs_for_generation` for both `GPT2DoubleHeadsModelOutput` and `GPT2LMHeadModel` as it was before - there is 0 problem with that. 

Note that we don't mind really if we have code duplication - that's one of the main philosophies of Transformers. We have re-implement the attention layer in 50+ models and this has worked well so far. 
It's far more important that we stay consistent with our APIs. `prepare_inptus_for_generation` is never implemented in the `...PretrainedModel` class, but always in the `...LMHeadModel` or `...ForCausalLM` class of each model type -> we should stick to this to not surprise a reader with a new code API.

patrickvonplaten commented on src/transformers/models/gpt2/modeling_gpt2.py:
```diff
@@ -574,6 +663,12 @@ class GPT2DoubleHeadsModelOutput(ModelOutput):
             - 0 for tokens that are **masked**.
 
             `What are attention masks? <../glossary.html#attention-mask>`__
+        attention_block_mask (:obj:`torch.BoolTensor` of shape or broadcastable to :obj:`(batch_size, num_heads, sequence_length, sequence_length)`, `optional`):
```
Maybe add a sentence that says "if None -> then `attention_block_mask` defaults to the `causal mask`

thomasw21 committed:
default to use signature default

Co-authored-by: Patrick von Platen <patrick.v.platen@gmail.com>

thomasw21 committed:
- Update docstring to explicit that default behaviour is causal masking
- Rename some lm_attention_mask to attention_block_mask that were missed previously

thomasw21 committed:
Woops

thomasw21 committed:
Test torch.allclose with False as expected result only on non empty tensors

thomasw21 commented:
Awaiting some comments from @patrickvonplaten .

thomasw21 commented:
un-stale

patrickvonplaten commented:
This week is the week haha - sorry to be so incredibly slow with this
